zcash_client_sqlite/src/wallet.rs

```diff
@@ -761,6 +761,54 @@ pub(crate) fn get_next_available_address<P: consensus::Parameters, C: Clock>(
     Ok(Some((addr, diversifier_index)))
 }
 
+pub(crate) fn list_addresses<P: consensus::Parameters>(
+    conn: &rusqlite::Connection,
+    params: &P,
+    account_uuid: AccountUuid,
+) -> Result<Vec<AddressInfo>, SqliteClientError> {
+    let mut addrs = vec![];
+
+    let mut stmt_addrs = conn.prepare(
+        "SELECT address, diversifier_index_be, key_scope
+         FROM addresses
+         JOIN accounts ON accounts.id = addresses.account_id
+         WHERE accounts.uuid = :account_uuid
+         AND exposed_at_height IS NOT NULL
+         ORDER BY exposed_at_height ASC, diversifier_index_be ASC",
+    )?;
+
+    let mut rows = stmt_addrs.query(named_params![
+        ":account_uuid": account_uuid.0,
+    ])?;
+
+    while let Some(row) = rows.next()? {
+        let addr_str: String = row.get(0)?;
+        let di_vec: Vec<u8> = row.get(1)?;
+        let _scope = KeyScope::decode(row.get(2)?)?;
+
+        let addr = Address::decode(params, &addr_str).ok_or_else(|| {
+            SqliteClientError::CorruptedData("Not a valid Zcash recipient address".to_owned())
+        })?;
+        let diversifier_index = decode_diversifier_index_be(&di_vec)?;
```

Comment:
Nonblocking/not significantly related to this PR: something that occurs to (and annoys) me is that when we start to do integration of zcashd wallet addresses that were generated from randomness, we'll no longer be able to maintain the not-null constraint on the diversifier index.

Reply:
Yep, I raised that point here: https://github.com/zcash/wallet/pull/33#discussion_r1995488331

There's a question to answer of whether or not we support migrating purely-random addresses or whether we require wallets to migrate away from them.

Reply:
I don't think we can require wallets to migrate away: a very common usage pattern by exchanges is to assign each user a distinct transparent address, and credit deposits to that address to that user. These addresses are then very often deeply embedded in end-user workflows: they may be whitelisted for withdrawals, embedded in scripts or internet point-of-sale system configuration, and so forth. These addresses will continue to receive funds so long as the outputs to the transparent protocol are allowed on the network, and Zallet must be able to spend those funds.